[dynamo] re-add option to log all guard check fails

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113585

Followup to https://github.com/pytorch/pytorch/pull/110325 - re-add the `report_all_guard_failures config` as a logging artifact `recompiles_verbose` with the following changes:
- evaluating the check must be wrapped with exception handling because subsequent code parts following the first failure may result in errors if evaluated (e.g. if a guard checks first for size, then tries to index - a guard failure due to insufficient size would result in an index error for the latter check).
- Adding a test for this case

Sample:
```python
import torch

def fn(x):
    return torch.rand(x[-1], len(x))

opt_fn = torch.compile(fn)
opt_fn([4, 5, 6])
opt_fn([7, 8])
opt_fn([9])
```

Output (with `TORCH_LOGS="recompiles_verbose"`):
```bash
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     triggered by the following guard failure(s):
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 0 failures:
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 3
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][0] == 4
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][1] == 5
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     triggered by the following guard failure(s):
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 0 failures:
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 2
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG] 
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 1 failures:
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 3
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][0] == 4
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

williamwen42 committed:
[dynamo] re-add option to log all guard check fails

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] re-add option to log all guard check fails"


Followup to https://github.com/pytorch/pytorch/pull/110927 - re-add the `report_all_guard_failures config` as `report_all_guard_failure_parts` with the following changes:
- evaluating the check must be wrapped with exception handling because subsequent code parts following the first failure may result in errors if evaluated (e.g. if a guard checks first for size, then tries to index - a guard failure due to insufficient size would result in an index error for the latter check).
- Adding a test for this case

Sample:
```python
import torch

def fn(x):
    return torch.rand(x[-1], len(x))

opt_fn = torch.compile(fn)
opt_fn([4, 5, 6])
opt_fn([7, 8])
opt_fn([9])
```

Output (with `report_all_guard_failures_config` true):
```bash
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416)
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 2
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - Ne(L['x'][1], 1)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG] 
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     guard 1 failures:
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jon-chuang commented on torch/_dynamo/config.py:
```diff
@@ -267,6 +267,9 @@
 report_guard_failures = True
 
+# [@compile_ignored: debug] For each guard failure, whether to print all failed parts, or only the first one
+report_all_guard_failure_parts = False
```
maybe we can call it `report_all_guard_failure_checks`

comment: `For each guard failure, whether to print all the failed checks of the guard, or only the first one`

jon-chuang reviewed (approved):
LGTM in principle, as long as tests pass. Thanks!

williamwen42 committed:
Update on "[dynamo] re-add option to log all guard check fails"


Followup to https://github.com/pytorch/pytorch/pull/110927 - re-add the `report_all_guard_failures config` as `report_all_guard_failure_parts` with the following changes:
- evaluating the check must be wrapped with exception handling because subsequent code parts following the first failure may result in errors if evaluated (e.g. if a guard checks first for size, then tries to index - a guard failure due to insufficient size would result in an index error for the latter check).
- Adding a test for this case

Sample:
```python
import torch

def fn(x):
    return torch.rand(x[-1], len(x))

opt_fn = torch.compile(fn)
opt_fn([4, 5, 6])
opt_fn([7, 8])
opt_fn([9])
```

Output (with `report_all_guard_failures_config` true):
```bash
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416)
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 2
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - Ne(L['x'][1], 1)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG] 
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     guard 1 failures:
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] re-add option to log all guard check fails"


Followup to https://github.com/pytorch/pytorch/pull/110927 - re-add the `report_all_guard_failures config` as `report_all_guard_failure_parts` with the following changes:
- evaluating the check must be wrapped with exception handling because subsequent code parts following the first failure may result in errors if evaluated (e.g. if a guard checks first for size, then tries to index - a guard failure due to insufficient size would result in an index error for the latter check).
- Adding a test for this case

Sample:
```python
import torch

def fn(x):
    return torch.rand(x[-1], len(x))

opt_fn = torch.compile(fn)
opt_fn([4, 5, 6])
opt_fn([7, 8])
opt_fn([9])
```

Output (with `report_all_guard_failures_config` true):
```bash
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416)
[2023-11-13 13:35:54,597] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 2
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - Ne(L['x'][1], 1)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG] 
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     guard 1 failures:
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416)
[2023-11-13 13:35:54,827] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 commented:
@pytorchbot merge

jon-chuang commented:
@williamwen42 need approval from core maintainer.

cc @ezyang as reviewer of previous related.

williamwen42 commented:
@jon-chuang do you think it would be better if I included an exception for guard checks that error out? e.g. in the example above, did the guard check `L['x'][2] == 6` fail because of an index error or because equality was not satisfied?

jon-chuang commented:
> do you think it would be better if I included an exception for guard checks that error out

I think it's better to print the exception as it's a verbose log? Just not the stacktrace.

williamwen42 committed:
Update on "[dynamo] re-add option to log all guard check fails"


Followup to https://github.com/pytorch/pytorch/pull/110927 - re-add the `report_all_guard_failures config` as `report_all_guard_failure_parts` with the following changes:
- evaluating the check must be wrapped with exception handling because subsequent code parts following the first failure may result in errors if evaluated (e.g. if a guard checks first for size, then tries to index - a guard failure due to insufficient size would result in an index error for the latter check).
- Adding a test for this case

Sample:
```python
import torch

def fn(x):
    return torch.rand(x[-1], len(x))

opt_fn = torch.compile(fn)
opt_fn([4, 5, 6])
opt_fn([7, 8])
opt_fn([9])
```

Output (with `report_all_guard_failures_config` true):
```bash
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416) (IndexError('list index out of range'))
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6 (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 2
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416) (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - Ne(L['x'][1], 1) (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2 (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2 (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG] 
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     guard 1 failures:
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416) (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5 (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416) (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6 (IndexError('list index out of range'))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented:
TBH, I wouldn't print the guards that fail with exceptions at all. They are guaranteed to have been adequately covered by earlier guards and are not contributing much.

ezyang commented on torch/_dynamo/config.py:
```diff
@@ -267,6 +267,11 @@
+# For each guard failure, whether to print all failed checks, or only the first one
+# (note: dynamo at runtime does not run past the first failed check)
+# [@compile_ignored: debug]
+report_all_guard_failure_checks = False
```
I think this should not be a config. Instead, it should be a logging artifact, so it can be operated remotely via TORCH_LOGS

jon-chuang replied:
hmm, I guess instead of `report_all_guard_failure_checks` one could check `all_guard_failures_log.isEnabledFor(logging.DEBUG)`?

williamwen42 commented:
@jon-chuang do you have an example where debugging would be easier if one knows which guard checks will fail after the first failure?

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -1249,36 +1249,43 @@ def get_guard_fail_reason(
+                fail_reason = eval(part, global_scope, scope)
+            except Exception as e:
+                if config.report_all_guard_failure_checks:
+                    fail_reason = f"{part} ({repr(e)})"
```
see my comment

ezyang reviewed (approved):
approving in the spirit of not getting bogged down

jon-chuang commented:
I don't have one off the top of my head.

But sometimes the order of the checks matter (e.g. a len check might fail, but there is some related other check about object identity that is more important to why the guard actually failed.

ezyang commented:
Sorry, to be clear, I'm OK with reporting all boolean expressions which return False.

But if I have `len(xs) == 2 and foo(xs[1]) and bar(xs[1])`, if xs is actually length 1, xs[1] is going to raise an exception, and I DON'T care about these guard failures, because they like don't matter, the whole point is that xs[1] doesn't exist, don't give me excruciating detail about all the guards on xs[1] when the element doesn't exist in the first place.

williamwen42 committed:
Update on "[dynamo] re-add option to log all guard check fails"


Followup to https://github.com/pytorch/pytorch/pull/110325 - re-add the `report_all_guard_failures config` as `report_all_guard_failure_parts` with the following changes:
- evaluating the check must be wrapped with exception handling because subsequent code parts following the first failure may result in errors if evaluated (e.g. if a guard checks first for size, then tries to index - a guard failure due to insufficient size would result in an index error for the latter check).
- Adding a test for this case

Sample:
```python
import torch

def fn(x):
    return torch.rand(x[-1], len(x))

opt_fn = torch.compile(fn)
opt_fn([4, 5, 6])
opt_fn([7, 8])
opt_fn([9])
```

Output (with `report_all_guard_failures_config` true):
```bash
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416) (IndexError('list index out of range'))
[2023-11-14 16:12:25,260] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6 (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     triggered by the following guard failure(s):
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     guard 0 failures:
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 2
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416) (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - Ne(L['x'][1], 1) (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2 (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - 2*L['x'][1] >= 2 (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG] 
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     guard 1 failures:
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - len(L['x']) == 3
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][0] == 4
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][1], 7640416) (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][1] == 5 (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - ___check_type_id(L['x'][2], 7640416) (IndexError('list index out of range'))
[2023-11-14 16:12:25,489] torch._dynamo.guards.__recompiles: [DEBUG]     - L['x'][2] == 6 (IndexError('list index out of range'))
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 commented:
Changes:
- use logging artifact `recompiles_verbose` instead of config
- do not output guard checks that error out

jon-chuang commented on torch/_dynamo/guards.py:
```diff
@@ -1236,6 +1238,14 @@ def replace(expr: str) -> Tuple[List[str], str]:
     return guard_body.getvalue(), make_guard_fn.getvalue()
 
 
+def is_recompiles_enabled():
```
You can just write `recompiles_verbose_log.isEnabledFor(logging.DEBUG)`

jon-chuang replied:
The log is only enabled if the artifact logger is enabled I believe?

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -1300,20 +1317,34 @@ def get_and_maybe_log_recompilation_reason(
         )
         if do_recompiles_log:
-            recompiles_log.debug(message, stack_info=True)
+            if recompiles_log.isEnabledFor(logging.DEBUG):
```
This is kind of weird, why are you using an if statement on isEnabledFor on recompiles_log here?

jon-chuang replied:
I think it should all go to verbose if verbose is enabled, and go to recompile if it is enabled. I.e. verbose will override the recompiles.

jon-chuang replied:
So verbose should be checked first, no?

williamwen42 replied:
I think it would be more complicated to log partially to regular `recompiles` and partially to `recompiles_verbose` - so I'm just picking one of them based on what's enabled.

ezyang replied:
yes

williamwen42 committed:
Update on "[dynamo] re-add option to log all guard check fails"


Followup to https://github.com/pytorch/pytorch/pull/110325 - re-add the `report_all_guard_failures config` as a logging artifact `recompiles_verbose` with the following changes:
- evaluating the check must be wrapped with exception handling because subsequent code parts following the first failure may result in errors if evaluated (e.g. if a guard checks first for size, then tries to index - a guard failure due to insufficient size would result in an index error for the latter check).
- Adding a test for this case

Sample:
```python
import torch

def fn(x):
    return torch.rand(x[-1], len(x))

opt_fn = torch.compile(fn)
opt_fn([4, 5, 6])
opt_fn([7, 8])
opt_fn([9])
```

Output (with `TORCH_LOGS="recompiles_verbose"`):
```bash
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     triggered by the following guard failure(s):
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 0 failures:
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 3
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][0] == 4
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][1] == 5
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     triggered by the following guard failure(s):
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 0 failures:
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 2
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG] 
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 1 failures:
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 3
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][0] == 4
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 commented:
- Only check if artifact is enabled, no need to check for the level of the logger object (enabled artifact assumes logger's level is proper)
- Prioritize logging to `recompiles_verbose` over `recompiles` if both are active.
- Fix bug in torch/_dynamo/utils.py: `logging.get_loggers() -> torch._logging.get_loggers()`

williamwen42 committed:
Update on "[dynamo] re-add option to log all guard check fails"


Followup to https://github.com/pytorch/pytorch/pull/110325 - re-add the `report_all_guard_failures config` as a logging artifact `recompiles_verbose` with the following changes:
- evaluating the check must be wrapped with exception handling because subsequent code parts following the first failure may result in errors if evaluated (e.g. if a guard checks first for size, then tries to index - a guard failure due to insufficient size would result in an index error for the latter check).
- Adding a test for this case

Sample:
```python
import torch

def fn(x):
    return torch.rand(x[-1], len(x))

opt_fn = torch.compile(fn)
opt_fn([4, 5, 6])
opt_fn([7, 8])
opt_fn([9])
```

Output (with `TORCH_LOGS="recompiles_verbose"`):
```bash
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     triggered by the following guard failure(s):
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 0 failures:
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 3
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][0] == 4
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][1] == 5
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     triggered by the following guard failure(s):
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 0 failures:
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 2
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG] 
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 1 failures:
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 3
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][0] == 4
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

williamwen42 commented:
@pytorchbot merge
